File: mirea/contests/2021-09-26 MIREA Welcome Contest 2021/NoPoint Python/Задача I Преподсчёт.py
MAX_A = 10 ** 4
SHIFTS = (2, 3, 5)
MOVES = len(SHIFTS)

dp = [[[] for __ in range(MAX_A)] for _ in range(MOVES)]


def precalc():
  for row in range(1, MAX_A + 1):
    cols = MAX_A // row
    for sub in dp:
      sub[row - 1].extend([False] * cols)

  for i in range(MAX_A):
    for j in range(len(dp[0][i])):
      for k in range(MOVES):
        prev_i = i - SHIFTS[k]
        if prev_i >= 0 and dp[(k + 1) % MOVES][prev_i][j] is False:
          dp[k][i][j] = True
        prev_j = j - SHIFTS[k]
        if prev_j >= 0 and dp[(k + 1) % MOVES][i][prev_j] is False:
          dp[k][i][j] = True


def main():
  precalc()

  tests = int(input())
  for test in range(tests):
    n, m = map(int, input().split())
    print("Alice" if dp[0][n - 1][m - 1] else "Bob")


main()

File: mirea/contests/2021-09-26 MIREA Welcome Contest 2021/NoPoint C++/Задача I Преподсчёт.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
#define int ll

vector<vector<vector<bool>>> precalc(int max_prod) {
  const vector<int> SHIFTS = {2, 3, 5};
  const int MOVES = (int)SHIFTS.size();
  vector<vector<vector<bool>>> dp(MOVES, vector<vector<bool>>(max_prod));
  for(int i = 0; i < max_prod; ++i) {
    int cols = max_prod / (i + 1);
    for(int k = 0; k < MOVES; ++k) {
      dp[k][i].resize(cols, false);
    }
  }

  for(int i = 0; i < max_prod; ++i) {
    for(int j = 0; j < (int)dp[0][i].size(); ++j) {
      for(int k = 0; k < MOVES; ++k) {
        int prev_i = i - SHIFTS[k];
        if(prev_i >= 0 && dp[(k + 1) % MOVES][prev_i][j] == false)
          dp[k][i][j] = true;
        int prev_j = j - SHIFTS[k];
        if(prev_j >= 0 && dp[(k + 1) % MOVES][i][prev_j] == false)
          dp[k][i][j] = true;
      }
    }
  }

  return dp;
}

int32_t main() {
  int queries;
  cin >> queries;
  vector<pair<int, int>> tests(queries);
  int max_prod = 0;
  for(int query = 0; query < queries; ++query) {
    auto &[n, m] = tests[query];
    cin >> n >> m;
    max_prod = max(max_prod, n * m);
  }
  vector<vector<vector<bool>>> dp = precalc(max_prod);
  for(auto &[n, m] : tests) {
    cout << (dp[0][n - 1][m - 1] ? "Alice" : "Bob") << endl;
  }
}
